relay-server/src/actors/global_config.rs

```diff
@@ -0,0 +1,121 @@
+use crate::actors::upstream::{SendQuery, UpstreamRelay};
+
+/// Service implementing the [`GlobalConfiguration`] interface.
+///
+/// The service is responsible for fetching the global config and
+/// forwarding it to the services that require it, and for serving downstream relays.
+#[derive(Debug)]
+pub struct GlobalConfigurationService {
+    global_config: Arc<GlobalConfig>,
+    envelope_processor: Addr<EnvelopeProcessor>,
+    upstream: Addr<UpstreamRelay>,
+}
+
+/// Service interface for the [`GetGlobalConfig`] message.
+pub enum GlobalConfiguration {
+    GetGlobalConfig(Sender<Arc<GlobalConfig>>),
+}
+
+impl Interface for GlobalConfiguration {}
+
+/// The message for requesting the most recent global config from [`GlobalConfigurationService`].
+pub struct GetGlobalConfig;
+
+impl FromMessage<GetGlobalConfig> for GlobalConfiguration {
+    type Response = AsyncResponse<Arc<GlobalConfig>>;
+
+    fn from_message(_: GetGlobalConfig, sender: Sender<Arc<GlobalConfig>>) -> Self {
+        Self::GetGlobalConfig(sender)
+    }
+}
+
+impl GlobalConfigurationService {
+    /// Creates a new [`GlobalConfigurationService`].
+    pub fn new(envelope_processor: Addr<EnvelopeProcessor>, upstream: Addr<UpstreamRelay>) -> Self {
+        let global_config = Arc::new(GlobalConfig::default());
+        Self {
+            global_config,
+            envelope_processor,
+            upstream,
+        }
+    }
+
+    fn handle_message(&self, message: GlobalConfiguration) {
+        match message {
+            GlobalConfiguration::GetGlobalConfig(sender) => {
+                sender.send(Arc::clone(&self.global_config));
+            }
+        }
+    }
+
+    /// Forwards the given global config to the services that require it.
+    fn update_global_config(&mut self, global_tx: UnboundedSender<Arc<GlobalConfig>>) {
+        let upstream_relay: Addr<UpstreamRelay> = self.upstream.clone();
+        let envelope_processor = self.envelope_processor.clone();
+
+        tokio::spawn(async move {
+            let query = GetProjectStates {
+                public_keys: vec![],
+                full_config: false,
+                no_cache: false,
+                global_config: true,
+            };
+
+            match upstream_relay.send(SendQuery(query)).await {
+                Ok(Ok(response)) => match response.global {
+                    Some(global_config) => {
+                        let global_config = Arc::new(global_config);
+                        if let Err(e) = global_tx.send(global_config.clone()) {
+                            relay_log::error!(
+                                "Failed to send global config to GlobalConfigurationService: {}",
+                                e
+                            );
+                        };
+                        envelope_processor.send::<Arc<GlobalConfig>>(global_config);
+                    }
+                    None => relay_log::error!("Upstream response didn't include a global config"),
```

Comment:
I would use something like following:
```suggestion
                    None => relay_log::error!("Global config is missing in upstream response"),
```

just to avoid the `didn't` with single quote in there.